### PR TITLE
ARROW-13373: [C++][Gandiva] Implement CRC32 Hive function on Gandiva

### PR DESCRIPTION
**Pull request by jpedroantunes**
Implement CRC32 Hive function on Gandiva

### CODE REVIEW

File: cpp/src/gandiva/crc32.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "gandiva/crc32.h"
+
+#include <boost/crc.hpp>
+
+namespace gandiva {
+namespace internal {
+// Get CRC32 for a given input chars
+int64_t GetCrc32(const char* input, int32_t input_len) {
+  boost::crc_32_type result;
```
Comment (projjal) [inline cpp/src/gandiva/crc32.cc:31]: Can't you add this snippet directly inside the function in gdv_function_stubs?

**Comment by projjal** [comment]
@ViniciusSouzaRoque Can you rebase this. There are conflicts

**Review by projjal — APPROVED** [review]

**Comment by kou** [comment]
This breaks our "java-jars" nightly build: https://github.com/ursacomputing/crossbow/runs/4543838221?check_suite_focus=true

```text
/usr/local/bin/ccache /opt/rh/devtoolset-10/root/usr/bin/c++ -DARROW_HAVE_RUNTIME_AVX2 -DARROW_HAVE_RUNTIME_AVX512 -DARROW_HAVE_RUNTIME_BMI2 -DARROW_HAVE_RUNTIME_SSE4_2 -DARROW_HAVE_SSE4_2 -DARROW_JEMALLOC -DARROW_JEMALLOC_INCLUDE_DIR="" -DARROW_WITH_RE2 -DARROW_WITH_TIMING_TESTS -DARROW_WITH_UTF8PROC -DGANDIVA_EXPORTING -DGANDIVA_LLVM_VERSION=12 -DUTF8PROC_STATIC -I/build/src -I/arrow/cpp/src -I/arrow/cpp/src/generated -isystem /arrow/cpp/thirdparty/flatbuffers/include -isystem /opt/vcpkg/installed/amd64-linux-static-release/include -isystem /build/jemalloc_ep-prefix/src -isystem /build/xsimd_ep/src/xsimd_ep-install/include -isystem /arrow/cpp/thirdparty/hadoop/include -Wno-noexcept-type -Wno-subobject-linkage  -fdiagnostics-color=always -O3 -DNDEBUG  -Wall -fno-semantic-interposition -msse4.2  -O3 -DNDEBUG -fPIC -std=c++14 -MD -MT src/gandiva/CMakeFiles/gandiva_objlib.dir/Unity/unity_3_cxx.cxx.o -MF src/gandiva/CMakeFiles/gandiva_objlib.dir/Unity/unity_3_cxx.cxx.o.d -o src/gandiva/CMakeFiles/gandiva_objlib.dir/Unity/unity_3_cxx.cxx.o -c /build/src/gandiva/CMakeFiles/gandiva_objlib.dir/Unity/unity_3_cxx.cxx
In file included from /build/src/gandiva/CMakeFiles/gandiva_objlib.dir/Unity/unity_3_cxx.cxx:9:
/arrow/cpp/src/gandiva/gdv_function_stubs.cc:22:10: fatal error: boost/crc.hpp: No such file or directory
   22 | #include <boost/crc.hpp>
      |          ^~~~~~~~~~~~~~~
compilation terminated.
```

Could you also update `ci/docker/java-jni-manylinux-201x.dockerfile`?

```diff
diff --git a/ci/docker/java-jni-manylinux-201x.dockerfile b/ci/docker/java-jni-manylinux-201x.dockerfile
index 021dab686..d30e6d341 100644
--- a/ci/docker/java-jni-manylinux-201x.dockerfile
+++ b/ci/docker/java-jni-manylinux-201x.dockerfile
@@ -20,16 +20,17 @@ FROM ${base}
 
 # Install the libaries required by the Gandiva to run
 RUN vcpkg install --clean-after-build \
-        llvm \
-        boost-system \
-        boost-date-time \
-        boost-regex \
-        boost-predef \
         boost-algorithm \
-        boost-locale \
+        boost-crc \
+        boost-date-time \
         boost-format \
-        boost-variant \
+        boost-locale \
         boost-multiprecision
+        boost-predef \
+        boost-regex \
+        boost-system \
+        boost-variant \
+        llvm
 
 # Install Java
 ARG java=1.8.0
```

@nealrichardson We may need to update our trimmed Boost archive.

**Comment by projjal** [comment]
@anthonylouisbsb Can you update `ci/docker/java-jni-manylinux-201x.dockerfile`

**Comment by anthonylouisbsb** [comment]
@projjal I will do it now

**Comment by anthonylouisbsb** [comment]
@kou https://github.com/apache/arrow/pull/11987 I opened a Pull Request to fix that problem. I am running the nightly builds right now.

**Pull request closed without merging**